### PR TITLE
Authentication with Passport and csrf (lesson13)

### PR DESCRIPTION
**Pull request by psharma1984**

### CODE REVIEW

File: controllers/sessionController.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
+const User = require("../models/User");
+const parseVErr = require("../util/parseValidationErr");
+const csrf = require('host-csrf');
+
+const registerShow = (req, res) => {
+    res.render("register");
+};
+
+const registerDo = async (req, res, next) => {
```
Comment (olearyma) [inline controllers/sessionController.js:9]: The provided code in the lesson had some errors with the registration validation.


For more information see #ferret-announcements in Slack
This updated code should fix the issues.
```
const registerDo = async (req, res, next) => {
  let validation_errors = false
  if (req.body.password != req.body.password1) {
    req.flash("error", "The passwords entered do not match.");
    validation_errors = true
  }

  try {
    await User.create(req.body);
  } catch (e) {
    if (e.constructor.name === "ValidationError") {
      parse_v(e, req);
    } else if (e.name === "MongoServerError" && e.code === 11000) {
      req.flash("error", "That email address is already registered.");
    } else {
      return next(e);
    }
    validation_errors = true
  }
  
  // if there are no validation errors then redirect to main paige 
  // else pass errors to be rendered in the register view

  if (! validation_errors)
  {
    res.redirect("/");
  }
  else
  {
    return res.render("register", {errors: req.flash("error")})
  }


};

```

Comment (psharma1984) [reply controllers/sessionController.js:9]: @olearyma Thanks for the fix :) . I updated the code.

File: controllers/sessionController.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    let validation_errors = false
+    if (req.body.password != req.body.password1) {
+        req.flash("error", "The passwords entered do not match.");
+        validation_errors = true;
+    }
+    try {
+        await User.create(req.body);
+    } catch (e) {
+        if (e.constructor.name === "ValidationError") {
+            parse_v(e, req);
```
Comment (olearyma) [inline controllers/sessionController.js]: Need to use the same name as  the const you define at top:
            parseVErr(e, req);

File: routes/sessionRoutes.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+} = require("../controllers/sessionController");
+
+router.route("/register").get(registerShow).post(registerDo);
+router
+  .route("/logon")
+  .get(logonShow)
+  .post(
+    passport.authenticate("local", {
+      successRedirect: "/",
+      failureRedirect: "sessions/logon",
```
Comment (olearyma) [inline routes/sessionRoutes.js]: Missing leading "/", it should be /sessions/logon

Current code causes a 404 error:
Sorry, /sessions/sessions/logon not found!

**Review by olearyma — APPROVED** [review]
Looks good.

**Pull request merged**